internal/inventory/tagprocessor.go

```diff
@@ -0,0 +1,364 @@
+			if rt.LastSeenUTC != nil && int64(*rt.LastSeenUTC) > lastSeenMicros {
+				lastSeenMicros = int64(*rt.LastSeenUTC)
+			}
+		}
+		if lastSeenMicros > 0 {
+			// divide originNanos by 1000 to get to micros
+			info.offsetMicros = (info.OriginNanos / 1000) - lastSeenMicros
+		}
+	}
+
+	for _, rt := range r.TagReportData {
+		if event := tp.processData(&rt, info); event != nil {
+			events = append(events, event)
+		}
+	}
+	return events, tp.snapshot()
+}
+
+// ReportInfo holds both pre-existing as well as computed metadata about an incoming ROAccessReport
+type ReportInfo struct {
+	DeviceName  string
+	OriginNanos int64
+
+	offsetMicros int64
+	// referenceTimestamp is the same as OriginNanos, but converted to milliseconds
+	referenceTimestamp int64
+}
+
+// NewReportInfo creates a new ReportInfo based on an EdgeX Reading value
+func NewReportInfo(reading *contract.Reading) ReportInfo {
+	return ReportInfo{
+		DeviceName:         reading.Device,
+		OriginNanos:        reading.Origin,
+		referenceTimestamp: reading.Origin / int64(time.Millisecond),
+	}
+}
+
+// getAlias returns the alias associated with a location if one has been defined,
+// otherwise it returns back the original location.
+func (tp *TagProcessor) getAlias(location string) string {
+	if alias, exists := tp.config.aliases[location]; exists && alias != "" {
+		return alias
+	}
+	return location
+}
+
+// Snapshot takes a snapshot of the entire tag inventory as a slice of StaticTag objects.
+// It does this by converting the inventory map of Tag pointers into a flat slice
+// of non-pointer StaticTags.
+func (tp *TagProcessor) snapshot() []StaticTag {
+	res := make([]StaticTag, 0, len(tp.inventory))
+	for _, tag := range tp.inventory {
+		staticTag := StaticTag{
+			EPC:           tag.EPC,
+			TID:           tag.TID,
+			Location:      tag.Location,
+			LocationAlias: tp.getAlias(tag.Location.String()),
+			LastRead:      tag.LastRead,
+			LastArrived:   tag.LastArrived,
+			LastDeparted:  tag.LastDeparted,
+			State:         tag.state,
+			StatsMap:      make(map[string]StaticTagStats, len(tag.statsMap)),
+		}
+
+		// re-populate the stats map
+		for loc, stats := range tag.statsMap {
+			if stats.rssiCount() == 0 {
+				continue // skip empty
+			}
+			staticTag.StatsMap[loc] = StaticTagStats{
+				LastRead: stats.LastRead,
+				MeanRSSI: stats.rssiDbm.Mean(),
+			}
+		}
+
+		res = append(res, staticTag)
+	}
+	return res
+}
+
+// processData processes an incoming TagReportData packet and updates the tag information and
+// device stats data structures.
+func (tp *TagProcessor) processData(rt *llrp.TagReportData, info ReportInfo) (event Event) {
+	var epc string
+	if len(rt.EPC96.EPC) > 0 {
+		epc = hex.EncodeToString(rt.EPC96.EPC)
+	} else {
+		epc = hex.EncodeToString(rt.EPCData.EPC)
+	}
+
+	tag, exists := tp.inventory[epc]
+	if !exists {
+		tag = NewTag(epc)
+		tp.inventory[epc] = tag
+	}
+	prevState, prevLoc := tag.state, tag.Location
+
+	defer func() {
```

Comment:
Comment why this needs to be deferred instead of just at the end of this function.